### ISSUE TITLE
build failure: multiple definition of `void Xpetra::Jacobi

### ISSUE THREAD

**Issue by nschloe** [trilinos/Trilinos#60]
The current master fails to build with
```
[...]
Wl,-Bdynamic ../../kokkos/algorithms/src/libtrilinos_kokkosalgorithms.so.12.5 ../../kokkos/core/src/libtrilinos_kokkoscore.so.12.5 
CMakeFiles/trilinos_stokhos_muelu.dir/sacado/kokkos/pce/muelu/MueLu_SaPFactory_UQ_PCE_Serial.cpp.o: In function `Xpetra::TpetraCrsGraph<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::getMap() const':
/«PKGBUILDDIR»/packages/stokhos/src/sacado/kokkos/pce/Sacado_UQ_PCE.hpp:145: multiple definition of `void Xpetra::Jacobi<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >(double, Xpetra::Vector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const&, Xpetra::Matrix<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const&, Xpetra::Matrix<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const&, Xpetra::Matrix<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >&, bool, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
```
Full details (including configuration) [at Launchpad](https://launchpadlibrarian.net/230424356/buildlog_ubuntu-xenial-amd64.trilinos_12.5~20151218030407-xenial1_BUILDING.txt.gz).

**Comment by etphipp** [comment]
@trilinos/xpetra 

@trilinos/muelu 

I see what the problem is.  Xpetra_IteratorOps.hpp defines a specialization of Xpetra::Jacobi in the header file:

#if defined(HAVE_XPETRA_EPETRA) && !defined(XPETRA_EPETRA_NO_32BIT_GLOBAL_INDICES)
  template<>
  void Jacobi<double,int,int,EpetraNode>(double omega,
                                         const Xpetra::Vector<double,int,int,EpetraNode> & Dinv,
                                         const Xpetra::Matrix<double,int,int,EpetraNode> & A,
                                         const Xpetra::Matrix<double,int,int,EpetraNode> & B,
                                         Xpetra::Matrix<double,int,int,EpetraNode> &C,
                                         bool call_FillComplete_on_result,
                                         bool doOptimizeStorage,
                                         const std::string & label) {

...

}

which will not work if there is ever more one instantiation of MueLu::SaPFactory.  This happens in Stokhos due to the additional instantiations on Stokhos' scalar types, but I believe would also arise in MueLu if it were instantiated on more than one node type.  The file needs to be changed to declare the specialization in the header file and put the definition in a cpp file.

Xpetra/MueLu devs:  do you ever test with more than one node instantiation?

-Eric

**Comment by aprokop** [comment]
This was fixed in the morning in 22372388.

@etphipp Yes, we do test with multiple nodes. This particular issue was not present for multiple nodes, but arose in one particular configuration for ETI.

**Comment by nschloe** [comment]
@aprokop Can you close the issue then?
